feat(bazel): (APF) Angular Package Format updates

alan-agius4 opened the pull request:
Several updates to Angular Package Format.

BREAKING CHANGE:

Several changes to the Angular Package Format (APF)
- Removal of FESM2015
- Replacing ES2020 with ES2022
- Replacing FESM2020 with FESM2022

itea-dev commented:
"Replacing FESM2020 with FESM2020" should be "Replacing FESM2020 with **FESM2022**"

alan-agius4 committed:
test: remove google-closure-compiler from tests

google-closure-compiler does not support ES2022

(cherry picked from commit 54365418bb8388ffa9b6388ae438e4d6ac4e081d)

alan-agius4 committed:
ci: disable system js tests

These are currently broken due to the APF v16 changes which dropped support for ES2015.

devversion commented on aio/content/examples/upgrade-phonecat-2-hybrid/rollup-config.js:
```diff
@@ -12,6 +12,9 @@ export default {
   },
+  // Rollup treeshaking has issues with https://github.com/angular/angular/blob/addd7f6249d54e258109f139fad1db0d0250352c/packages/core/src/linker/query_list.ts#L192
+  // see: https://github.com/rollup/rollup/issues/4895
+  treeshake: false,
```
The fix seems to be released already. Should we remove this again & update? (although you declared it as broken anyway)

alan-agius4 replied:
I did try to update rollup but Zone tests started to break. I'd prefer to update rollup on a separate PR especially since it's a major bump.

devversion replied:
sounds good. resolving this comment!

devversion commented on aio/content/examples/upgrade-phonecat-3-final/BUILD.bazel:
```diff
@@ -5,4 +5,5 @@ package(default_visibility = ["//visibility:public"])
 docs_example(
     name = "upgrade-phonecat-3-final",
     flaky = True,  # TODO: figure out why this is flaky or times out
+    test_tags = ["broken"],
```
We should add a comment on why we declare it as "broken" at places we added the tag.

devversion commented on aio/content/guide/angular-package-format.md:
```diff
@@ -440,7 +398,7 @@ Node.js has one that is well specified and widely used, TypeScript supports seve
 Common module formats are CommonJS \(CJS, typically used for Node.js applications\) or ECMAScript Modules \(ESM\).
 The module format indicates only the packaging of the individual modules, but not the JavaScript language features used to make up the module content.
-Because of this, the Angular team often uses the language level specifier as a suffix to the module format, \(for example, ESM+ES2015 specifies that the module is in ESM format and contains code down-leveled to ES2015\).
+Because of this, the Angular team often uses the language level specifier as a suffix to the module format, \(for example, ESM+ES202020 specifies that the module is in ESM format and contains ES2022 code\).
```
```suggestion
Because of this, the Angular team often uses the language level specifier as a suffix to the module format, \(for example, ESM+ES2022 specifies that the module is in ESM format and contains ES2022 code\).
```

alan-agius4 replied:
Good catch

devversion commented on packages/bazel/src/ng_package/rollup.config.js:
```diff
@@ -187,7 +163,7 @@ const plugins = [
   },
   nodeResolve({
-    mainFields: ['es2020', 'es2015', 'module', 'browser'],
+    mainFields: ['module', 'browser'],
```
We need to be careful here. Have you looked into this. There is a risk in bundling incorrect stuff with this change, as we might have relied on such a main field for dependencies of APF.

alan-agius4 replied:
Interesting, reverting.

devversion commented on packages/bazel/src/ng_module/ng_module.bzl:
```diff
@@ -202,13 +202,16 @@ def _ngc_tsconfig(ctx, files, srcs, **kwargs):
+    # TODO(alanagius): The below causes a drastic size increase when enabled. This is mainly due to Babel transforms for Safari 15.
+    # https://github.com/angular/angular-cli/blob/3e8bdf72d6b7e2925d2822da807b726f88a77e1a/packages/angular_devkit/build_angular/src/babel/presets/application.ts#L199-L204
+    tsconfig["compilerOptions"]["useDefineForClassFields"] = False
+    tsconfig["compilerOptions"]["target"] = "es2022"
```
We should do the same changes in `ts_library` via a patch? `ng_module` and `ts_library` should mostly have the same TS target and emit options

alan-agius4 replied:
That would be nice to do.

devversion replied:
We should look into doing it here to keep both `ng_module` and `ts_library` in sync. Not all targets shipped to npm are necessarily using `ng_module` and we keep both in sync as much as possible (in terms of TS target/options)

alan-agius4 replied:
Patched ts_library to be able to produce es2022. I did force benchpress to be compiler with es2020 though as it requires a some effort to make it work with es2022.

I’ll try to fix that in a follow up.

alxhub commented on packages/elements/src/component-factory-strategy.ts:
```diff
@@ -81,17 +81,21 @@ export class ComponentNgElementStrategy implements NgElementStrategy {
+  private readonly unchangedInputs: Set<string>;
 
   /** Service for setting zone context. */
-  private readonly ngZone = this.injector.get<NgZone>(NgZone);
```
What is the issue with initializing fields this way?

alan-agius4 replied:
As discussed offline this is due to the changes in ES2022 The problem there is that `ngZone` property has a dependency on a property in the constructor `this.injector` which is used before it's being initialised. 

Will update the usages to use the `inject` function.

alan-agius4 replied:
I tried to use the `inject` function, but in this case the class is not an injectable, as such the parameters in the constructor are needed so that the values are provided when creating an new instance.

Example: https://github.com/angular/angular/blob/92e41e9cb417223d9888a4c23b4c0e73188f87d0/packages/elements/src/component-factory-strategy.ts#L35

alan-agius4 committed:
feat(bazel): (APF) Angular Package Format updates

Several updates to Angular Package Format.

BREAKING CHANGE:

Several changes to the Angular Package Format (APF)
- Removal of FESM2015
- Replacing ES2020 with ES2022
- Replacing FESM2020 with FESM2022

alan-agius4 committed:
refactor: update code to be ES2022 compliant

This commit updates parts of the FW to be ES2022 complaint.

These changes are needed to fix the following problems problems with using properties before they are initialized.

Example
```ts
class Foo {
   bar = this.buz;
   constructor(private buz: unknown){}
}
```

alan-agius4 committed:
build: update `ts_library` to produce `ES2022`

This commit patches `ts_library` to be able to produce `ES2022`. Also, updates the build tsconfig and sets `useDefineForClassFields` to `false` to keep the same behaviour of `ng_module`.

alan-agius4 committed:
build: force benchpress to be compiled with ES2020

Benchpress requires a decent amount of work to be ES2022 compliant. This will be done in a followup PR.

alan-agius4 commented on yarn.lock:
```diff
@@ -136,7 +136,6 @@
 
```
Note to self. Delete this

alan-agius4 commented on aio/content/examples/upgrade-module/BUILD.bazel:
```diff
@@ -8,5 +8,9 @@ docs_example(
-    test_tags = ["requires-network"],
+    test_tags = [
+        "requires-network",
+        # This is currently broken due to usage of ESM2022 which does not work out-of-the-box with system.js.
```
I’ll create as a reminder, but I plan to look into this during the feature freeze.

devversion replied:
@alan-agius4 Did we ever create an issue for this?

alan-agius4 replied:
No, still in my todo list. I'll try to create one in the coming days.

AndrewKushnir reviewed (approved):
👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

(FYI the file affected public API: aio/content/guide/angular-package-format.md)

alxhub reviewed (approved):
Reviewed-for: public-api

atscott commented:
This PR was merged into the repository by commit 132c00c2f88faf97591de0477f80397e5e3c31d4.
